src/internet/subsonic/subsonicdynamicplaylist.cpp

```diff
@@ -0,0 +1,246 @@
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Clementine.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "subsonicdynamicplaylist.h"
+#include "subsonicservice.h"
+
+#include <QEventLoop>
+#include <QFileInfo>
+#include <QSslConfiguration>
+#include <QXmlStreamReader>
+
+#include "core/application.h"
+#include "core/logging.h"
+#include "core/network.h"
+#include "core/taskmanager.h"
+#include "core/timeconstants.h"
+#include "core/waitforsignal.h"
+#include "internet/core/internetplaylistitem.h"
+
+#include <boost/scope_exit.hpp>
+
+// 500 limit per subsonic api
+const int SubsonicDynamicPlaylist::kMaxCount = 500;
+const int SubsonicDynamicPlaylist::kDefaultCount = 10;
+const int SubsonicDynamicPlaylist::kDefaultOffset = 0;
+
+SubsonicDynamicPlaylist::SubsonicDynamicPlaylist()
+    : stat_(QueryStat_Newest), offset_(0) {}
+
+SubsonicDynamicPlaylist::SubsonicDynamicPlaylist(const QString& name,
+                                                 QueryStat stat)
+    : stat_(stat), offset_(0) {
+  set_name(name);
+}
+
+void SubsonicDynamicPlaylist::Load(const QByteArray& data) {
+  QDataStream s(data);
+  s >> *this;
+}
+
+void SubsonicDynamicPlaylist::Load(QueryStat stat) { stat_ = stat; }
+
+QByteArray SubsonicDynamicPlaylist::Save() const {
+  QByteArray ret;
+  QDataStream s(&ret, QIODevice::WriteOnly);
+  s << *this;
+  return ret;
+}
+
+// copied from SubsonicService
+QNetworkReply* SubsonicDynamicPlaylist::Send(QNetworkAccessManager& network,
+                                             const QUrl& url,
+                                             const bool usesslv3) {
+  QNetworkRequest request(url);
+  // Don't try and check the authenticity of the SSL certificate - it'll almost
+  // certainly be self-signed.
+  QSslConfiguration sslconfig = QSslConfiguration::defaultConfiguration();
+  sslconfig.setPeerVerifyMode(QSslSocket::VerifyNone);
+  if (usesslv3) {
+    sslconfig.setProtocol(QSsl::SslV3);
+  }
+  request.setSslConfiguration(sslconfig);
+  QNetworkReply* reply = network.get(request);
+  return reply;
+}
+
+PlaylistItemList SubsonicDynamicPlaylist::Generate() {
+  return GenerateMore(10);
+}
+
+PlaylistItemList SubsonicDynamicPlaylist::GenerateMore(int count) {
+  SubsonicService* service = InternetModel::Service<SubsonicService>();
+  const int task_id =
+      service->app_->task_manager()->StartTask(tr("Fetching Playlist Items"));
+
+  BOOST_SCOPE_EXIT((service)(task_id)) {
+    // stop task when we're done
+    service->app_->task_manager()->SetTaskFinished(task_id);
+  }
+  BOOST_SCOPE_EXIT_END
+
+  QUrl url = service->BuildRequestUrl("GetAlbumList");
+  QNetworkAccessManager network;
+
+  url.addQueryItem("type", GetTypeString());
+
+  if (count > kMaxCount) count = kMaxCount;
+  if (count != kDefaultCount) {
```

Comment:
Any reason not to just always set the `size` and `offset` parameters? It would make this code more robust to server changes.